### PR TITLE
Add support for adding partitions when flashing through tools/firmware

### PR DESCRIPTION
**Pull request by kasperl**

### CODE REVIEW

**Comment by kasperl** [comment]
```
tools/firmware -e firmware.envelope --port /dev/ttyUSB0 --partition file:foo=README.md --partition empty:bar=4096
```

File: tools/firmware.toit [hunk 1]
```diff
@@ -619,30 +632,74 @@ flash parsed/cli.Parsed -> none:
...
+  parsed_partitions.do: | entry/Map |
+    description := ?
+    is_file := entry.contains "file"
+    if is_file: description = entry["file"]
+    else: description = entry["empty"]
+    assign_index := description.index_of "="
+    if assign_index < 0: throw "malformed partition description '$description'"
+    name := description[..assign_index]
+    if not (0 < name.size <= 15): throw "malformed partition name '$name'"
+    if partitions.get name: throw "duplicate partition named '$name'"
```
Comment (floitsch) [inline tools/firmware.toit:650]:
```suggestion
    if partitions.contains name: throw "duplicate partition named '$name'"
```

File: tools/firmware.toit [hunk 2]
```diff
@@ -1223,3 +1280,39 @@ find_details_offset bits/ByteArray -> int:
   // No magic numbers were found so the image is from a legacy SDK that has the
   // image details at a fixed offset.
   throw "cannot find magic marker in binary file"
+
+// TODO(kasper): Move this to the cli package?
```
Comment (floitsch) [inline tools/firmware.toit:1284]: Agreed.

**Review by floitsch — APPROVED** [review]
LGTM.

**Pull request merged**